src/ipc/ipc4/logging.c

```diff
@@ -150,27 +154,47 @@ int ipc4_logging_enable_logs(bool first_block,
 	return 0;
 }
 
-int ipc4_logging_shutdown(void)
-{
-	struct ipc4_log_state_info log_state = { 0 };
-
-	/* log_state.enable set to 0 above */
-
-	return ipc4_logging_enable_logs(true, true, sizeof(log_state), (char *)&log_state);
-}
+#endif
 
-#else
+#ifdef CONFIG_LOG_BACKEND_SOF_PROBE
 int ipc4_logging_enable_logs(bool first_block,
 			     bool last_block,
 			     uint32_t data_offset_or_size,
 			     const char *data)
 {
-	return IPC4_UNKNOWN_MESSAGE_TYPE;
-}
+	const struct log_backend *log_backend = log_backend_probe_get();
+	const struct ipc4_log_state_info *log_state;
+
+	if (!(first_block && last_block))
+		return -EINVAL;
+
+	if (data_offset_or_size < sizeof(struct ipc4_log_state_info))
+		return -EINVAL;
+
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)data, data_offset_or_size);
+
+	log_state = (const struct ipc4_log_state_info *)data;
+
+	if (log_state->enable) {
```

Comment:
mtrace and probes has not been exclusive before, so we'd have to somehow make them work side-by-side after this PR as well. Could you rework this to have  ashared enable_logs handling and then have a helper function to actually implement "enable" and "disable" and these functions could have the ifdefs for mtrace and probe. Then you could at least build both, and in worst case, we can log bugs if the two cannot be enabled at the same time.

Reply:
Logging through probes at the same time as mtrace ? Technically possible I think but it is useless since You only ever use one method. Windows will use probes and linux will use mtrace hence the distinction.
Logging by probes only happens only  if the logging is configured in Enable Logs function. Probe logging is always ready to go, but only starts on enable logs witch will set probes as active.

The very idea of how it is done now with ifdefs rubs me the wrong way - this should not look like that. This area needs an overhaul. So if possible I would like to keep theme separate for now. This may in future help in a way (I had an idea to save some func pointers and overwrite theme case by case - but it needs more time in "head oven").
That's why I would like to push it as is since it enables logging for windows and does not impact anyone else.

Reply:
@dabekjakub We don't need to refactor the code now, agreed, but even if we do the minimal change, we can't break existing stuff. We have probes support in Linux as well, so we need to keep that working. I'm ok to make mtrace and probe exclusive  (at build time) but this needs to added as a ifdef check that raises a (clear) build error if both are selected.

Second, I don't still understand how this matches with "sof/src/probe.c:probe_point_add()" that also adds probe_logging_init(). So now you have two places in code where probe_logging_init() is called. I thought also Windows will use this API (this code was added by @abonislawski ). I mean somehow Windows also needs to set the DMA channel for probe, I'd assume it's a IPC4_PROBE_MODULE_PROBE_POINTS_ADD. 

If the two are exclusive, then I'm ok with the current code, but it looks strange we have two IPC4 handlers doing the same thing and with no code to handle concurrent use. Also, I don't see how this handles the case where ENABLE_LOGS is sent (with probes enabled), but PROBE_POINTS_ADD is not sent. You'd start calling the probe hook, but the DMA channel is not established, so it would seem the FW would crash in probe.c:probe_logging_hook() in this case. I know this the expected flow, but FW must not crash if sent IPCs in invalid sequence.

Reply:
I will test scenarios with probe adding on Monday - and will remove the duplicate calls.
Will add some assert for both enabled - this would be up for change in future.
However I disagree with FW being a flow controller. What would You check for ?  Crash should avoided if possible that is true but how would You handle it here ?

Reply:
Only then we have to follow the state of either DMA or flow. And FW should not ever handle the flow itself. The same situation would occur if You demanded playback at wrongly configurated gateway i think. Only way to do it semi correctly would be checking if hook was connected.

Reply:
all - can we make it best practice now to inline comment all flows around cache usage.  i.e. to make it obvious to the user why we are doing the cache operation and on what.